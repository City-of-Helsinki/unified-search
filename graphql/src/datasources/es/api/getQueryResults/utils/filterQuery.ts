import { DateTime } from 'luxon';
import { DEFAULT_TIME_ZONE } from '../../../constants';
import type {
  ArrayFilter,
  MustHaveReservableResourceFilter,
  OpenAtFilter,
} from '../../../types';
import type { QueryResultFilterProps } from '../types';
import { buildArrayFilter } from '../../../utils';

/**
 * @deprecated
 *
 * TODO: Reservable is still needed and should be re-implemented in https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-578.
 */
// const buildMustHaveReservableResourceFilter =
//   (): MustHaveReservableResourceFilter => ({
//     bool: {
//       should: [
//         // i.e. OR
//         {
//           term: {
//             'venue.resources.reservable': true,
//           },
//         },
//         {
//           exists: {
//             field: 'venue.resources.externalReservationUrl',
//           },
//         },
//       ],
//     },
//   });

export function getFilters({
  administrativeDivisionIds,
  ontologyTreeIdOrSets,
  ontologyWordIdOrSets,
  providerTypes,
  serviceOwnerTypes,
  targetGroups,
  mustHaveReservableResource,
  openAt,
}: QueryResultFilterProps): Array<
  ArrayFilter | OpenAtFilter | MustHaveReservableResourceFilter
> {
  // Assume time zone DEFAULT_TIME_ZONE when there is no time zone offset provided by the client.
  const openAtDateTime = DateTime.fromISO(openAt ?? '', {
    zone: DEFAULT_TIME_ZONE,
  });
  const finalOpenAt = openAtDateTime.isValid ? openAtDateTime.toISO() : openAt;

  return [
    ...buildArrayFilter(
      'venue.location.administrativeDivisions.id.keyword',
      administrativeDivisionIds ?? []
    ),
    ...(ontologyTreeIdOrSets ?? []).flatMap((ontologyTreeIdOrSet) =>
      buildArrayFilter(
        'links.raw_data.ontologytree_ids_enriched.id',
        ontologyTreeIdOrSet ?? []
      )
    ),
    ...(ontologyWordIdOrSets ?? []).flatMap((ontologyWordIdOrSet) =>
      buildArrayFilter(
        'links.raw_data.ontologyword_ids_enriched.id',
        ontologyWordIdOrSet ?? []
      )
    ),
    ...buildArrayFilter(
      'venue.serviceOwner.providerType.keyword',
      providerTypes ?? []
    ),
    ...buildArrayFilter(
      'venue.serviceOwner.type.keyword',
      serviceOwnerTypes ?? []
    ),
    ...buildArrayFilter('venue.targetGroups.keyword', targetGroups ?? []),
    // TODO: Reservable is still needed and should be re-implemented in https://helsinkisolutionoffice.atlassian.net/browse/LIIKUNTA-578.
    // ...(mustHaveReservableResource
    //   ? [buildMustHaveReservableResourceFilter()]
    //   : []),
    ...(finalOpenAt
      ? [
          {
            term: {
              'venue.openingHours.openRanges': finalOpenAt,
            },
          },
        ]
      : []),
  ];
}

export function filterQuery(query: any, filterProps: QueryResultFilterProps) {
  const filters = getFilters(filterProps);
  if (filters.length > 0) {
    query.query.bool.minimum_should_match = 1;
    query.query.bool.filter = filters;
  }
}
