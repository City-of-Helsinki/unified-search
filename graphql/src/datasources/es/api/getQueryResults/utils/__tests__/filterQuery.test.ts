import { describe, expect, it } from 'vitest';

import { filterQuery } from '../filterQuery.js';

describe('filterQuery', () => {
  it('test query mutation with all filters', () => {
    const origQuery = { query: { bool: { someFieldName: 'someFieldValue' } } };
    const query = structuredClone(origQuery);
    const filterProps = {
      administrativeDivisionIds: ['adminDivId1', 'adminDivId2'],
      ontologyTreeIdOrSets: [['treeId1'], ['treeId2', 'treeId3']],
      ontologyWordIdOrSets: [['wordId1', 'wordId2'], ['wordId3']],
      providerTypes: ['ASSOCIATION', 'SELF_PRODUCED'],
      serviceOwnerTypes: ['MUNICIPAL_SERVICE', 'PRIVATE_SERVICE'],
      targetGroups: ['CHILDREN_AND_FAMILIES', 'YOUTH'],
      mustHaveReservableResource: true,
      openAt: '2025-06-25T12:34:45.123+06:00',
    };
    filterQuery(query, filterProps);

    // Expect to mutate the input parameter `query`
    expect(query).not.toEqual(origQuery);

    // Check the overall structure of the generated query
    expect(Object.keys(query)).toStrictEqual(['query']);
    expect(Object.keys(query.query)).toStrictEqual(['bool']);
    expect(query.query.bool).toHaveProperty('minimum_should_match', 1);
    expect(query.query.bool).toHaveProperty('someFieldName', 'someFieldValue');
    expect(query.query.bool).toHaveProperty('filter');
    expect(Object.keys(query.query.bool)).toHaveLength(3);

    // Check the generated filters
    //
    // @ts-expect-error 'filter' does exist as expected above
    expect(query.query.bool.filter).toMatchSnapshot();
  });

  it('test non-op without filters', () => {
    const origQuery = {
      query: { bool: { someFieldName: 'someFieldValue' } },
    } as const;
    const query = structuredClone(origQuery);
    const filterProps = {};
    filterQuery(query, filterProps);

    // Expect the input parameter `query` to remain unchanged
    expect(query).toStrictEqual(origQuery);
  });
});
