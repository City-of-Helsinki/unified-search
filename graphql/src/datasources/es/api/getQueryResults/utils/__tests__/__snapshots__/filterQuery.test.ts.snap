// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filterQuery > test query mutation with all filters 1`] = `
[
  {
    "bool": {
      "should": [
        {
          "term": {
            "venue.location.administrativeDivisions.id.keyword": "adminDivId1",
          },
        },
        {
          "term": {
            "venue.location.administrativeDivisions.id.keyword": "adminDivId2",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "links.raw_data.ontologytree_ids_enriched.id": "treeId1",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "links.raw_data.ontologytree_ids_enriched.id": "treeId2",
          },
        },
        {
          "term": {
            "links.raw_data.ontologytree_ids_enriched.id": "treeId3",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "links.raw_data.ontologyword_ids_enriched.id": "wordId1",
          },
        },
        {
          "term": {
            "links.raw_data.ontologyword_ids_enriched.id": "wordId2",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "links.raw_data.ontologyword_ids_enriched.id": "wordId3",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "venue.serviceOwner.providerType.keyword": "ASSOCIATION",
          },
        },
        {
          "term": {
            "venue.serviceOwner.providerType.keyword": "SELF_PRODUCED",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "venue.serviceOwner.type.keyword": "MUNICIPAL_SERVICE",
          },
        },
        {
          "term": {
            "venue.serviceOwner.type.keyword": "PRIVATE_SERVICE",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "venue.targetGroups.keyword": "CHILDREN_AND_FAMILIES",
          },
        },
        {
          "term": {
            "venue.targetGroups.keyword": "YOUTH",
          },
        },
      ],
    },
  },
  {
    "bool": {
      "should": [
        {
          "term": {
            "venue.reservation.reservable": true,
          },
        },
        {
          "exists": {
            "field": "venue.reservation.externalReservationUrl",
          },
        },
      ],
    },
  },
  {
    "term": {
      "venue.openingHours.openRanges": "2025-06-25T09:34:45.123+03:00",
    },
  },
]
`;
